// @mintbase-js/sdk/lib/config/config.mocks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MAINNET_MOCK = exports.TESTNET_MOCK = void 0;
const types_1 = require("../types");
exports.TESTNET_MOCK = {
    apiKey: process.env.MINTBASE_API_KEY,
    contractAddress: 'ruben.mintspace2.testnet',
    callbackUrl: 'https://mintbase.xyz/success',
    debugMode: true,
    graphqlUrl: types_1.GRAPHQL_ENDPOINTS.testnet,
    isSet: true,
    marketAddress: types_1.MARKET_CONTRACT_ADDRESS.testnet,
    mbContract: types_1.MINTBASE_CONTRACTS.testnet,
    nearRpcUrl: types_1.RPC_ENDPOINTS.testnet,
    network: types_1.NEAR_NETWORKS.TESTNET
};
exports.MAINNET_MOCK = {
    apiKey: process.env.MINTBASE_API_KEY,
    contractAddress: 'zengin.mintbase1.near',
    callbackUrl: 'https://mintbase.xyz/success',
    debugMode: false,
    graphqlUrl: types_1.GRAPHQL_ENDPOINTS.mainnet,
    isSet: true,
    marketAddress: types_1.MARKET_CONTRACT_ADDRESS.mainnet,
    mbContract: types_1.MINTBASE_CONTRACTS.mainnet,
    nearRpcUrl: types_1.RPC_ENDPOINTS.mainnet,
    network: types_1.NEAR_NETWORKS.MAINNET
};
