// @mintbase-js/sdk/lib/removeMinter/removeMinter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.removeMinter = void 0;
const config_1 = require("../config/config");
const constants_1 = require("../constants");
const errorMessages_1 = require("../errorMessages");
const types_1 = require("../types");
/**
 * Remove minting access of a provided id for a contract you own.
 * @param removeMinterArgs {@link RemoveMinterArgs}
 * @returns contract call to be passed to @mintbase-js/sdk execute method
 */
const removeMinter = (args) => {
    const { minterId, contractAddress = config_1.mbjs.keys.contractAddress } = args;
    if (contractAddress == null) {
        throw new Error(errorMessages_1.ERROR_MESSAGES.CONTRACT_ADDRESS);
    }
    return {
        contractAddress: contractAddress || config_1.mbjs.keys.contractAddress,
        args: {
            account_id: minterId,
        },
        methodName: types_1.TOKEN_METHOD_NAMES.REMOVE_MINTER,
        gas: constants_1.GAS,
        deposit: constants_1.ONE_YOCTO,
    };
};
exports.removeMinter = removeMinter;
