// @mintbase-js/storage/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OBJECT_IS_EMPTY_ERROR = exports.ANON_USER_WARNING = exports.MINTBASE_API_ANON_USER = exports.MINTBASE_API_KEY_HEADER = exports.MINTBASE_API_KEY = exports.supportedStorageServices = exports.nodes = exports.MAX_UPLOAD_ERROR_MSG = exports.MAX_UPLOAD_BYTES = exports.ARWEAVE_SERVICE_HOST = void 0;
exports.ARWEAVE_SERVICE_HOST = 'https://ar.mintbase.xyz';
exports.MAX_UPLOAD_BYTES = 31457280;
exports.MAX_UPLOAD_ERROR_MSG = 'The max file size supported is currently is 30MB. Please contact Mintbase support regarding use cases for large files.';
exports.nodes = {
    bundlr: {
        node1: {
            url: 'http://node1.bundlr.network',
            supportedCurrencies: ['arweave'],
        },
        node2: {
            url: 'http://node2.bundlr.network',
            supportedCurrencies: ['arweave'],
        },
    },
};
exports.supportedStorageServices = {
    arweave: {
        bundlr: 'bundlr',
    },
};
exports.MINTBASE_API_KEY = 'anon';
exports.MINTBASE_API_KEY_HEADER = 'mb-api-key';
exports.MINTBASE_API_ANON_USER = 'anon';
exports.ANON_USER_WARNING = 'Warning: you are using the anonymous mintbase api key. You need to specify yours in order to upload a file. In the future this will be mandatory';
exports.OBJECT_IS_EMPTY_ERROR = 'Provided object is empty';
