// @mintbase-js/sdk/lib/burn/burn.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.burn = void 0;
const constants_1 = require("../constants");
const config_1 = require("../config/config");
const types_1 = require("../types");
const errorMessages_1 = require("../errorMessages");
/**
 * Burns one or more tokens from a given contract.
 * @param burnArguments {@link BurnArgs}
 * @returns contract call to be passed to @mintbase-js/sdk execute method
 */
const burn = ({ tokenIds, contractAddress = config_1.mbjs.keys.contractAddress }) => {
    if (contractAddress == null) {
        throw new Error(errorMessages_1.ERROR_MESSAGES.CONTRACT_ADDRESS);
    }
    if (tokenIds.length === 0) {
        throw new Error(errorMessages_1.ERROR_MESSAGES.BURN_TOKEN_IDS);
    }
    return {
        contractAddress: contractAddress || config_1.mbjs.keys.contractAddress,
        methodName: types_1.TOKEN_METHOD_NAMES.BATCH_BURN,
        args: {
            token_ids: tokenIds,
        },
        deposit: constants_1.DEPOSIT_FOR_BURN,
        gas: constants_1.GAS_FOR_BURN,
    };
};
exports.burn = burn;
